### PR TITLE
enhancement/15348-heikin-ashi

### PR DESCRIPTION
**Pull request by karolkolodziej**
Enhancement #15348, added Heikin Ashi series.

To do's:
- [x] - new series/options implementation
 - [x]  - ~~update [SCSS](https://github.com/highcharts/highcharts/blob/master/css/highcharts.scss) (in case of new styled-mode options) and the [list](https://github.com/highcharts/highcharts/blob/master/docs/chart-design-and-style/style-by-css.md) (if necessary)~~
- [x] - add to [Stock Tools'](https://www.highcharts.com/demo/stock/stock-tools-gui) button list of series types
- [x] - official [demo](https://www.highcharts.com/demo/stock) (this will cover visual test too)
- [x] - official [general documentation](https://www.highcharts.com/docs/stock/candlestick-chart)
- [x] - update [API](https://api.highcharts.com/highstock/) (series/options description)
- [x] - fix `heikinashi` with `colorAxis`
- [x] - TS
- [x] - add `dataGrouping` option
- [x] - position demo

### CODE REVIEW

File: docs/stock/heikinashi.md
Comment (TorsteinHonsi) [inline docs/stock/heikinashi.md:1]: Wrong header?

File: docs/stock/heikinashi.md
Comment (TorsteinHonsi) [inline docs/stock/heikinashi.md:5]: > A Heiken Ashi chart

Should we use a different word than "chart"? This is an item or a point, right?

.. _have_ been modified by the formula.

File: samples/stock/demo/heikinashi/demo.details
Comment (TorsteinHonsi) [inline samples/stock/demo/heikinashi/demo.details:2]: Wrong name?

File: samples/stock/demo/heikinashi/demo.js
Comment (TorsteinHonsi) [inline samples/stock/demo/heikinashi/demo.js:33]: It is not easy to see the difference between the candlestick and the Heikin Ashi without zooming in. I propose setting `rangeSelector.selected` to 1 in the config.

File: samples/stock/demo/heikinashi/demo.js
Comment (TorsteinHonsi) [inline samples/stock/demo/heikinashi/demo.js:14]: As far as I can see, this is not necessary. It just copies the data. We can just use `data` directly on both series.

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]

**Comment by TorsteinHonsi** [comment]
Thanks for the PR! Please see my inline comments.

File: samples/stock/demo/heikinashi/demo.js
Comment (pawellysy) [inline samples/stock/demo/heikinashi/demo.js:11]: `candlestick`?

**Review by pawellysy — CHANGES_REQUESTED** [review]
minor changes

File: ts/Series/HeikinAshi/HeikinAshiPoint.ts
Comment (raf18seb) [inline ts/Series/HeikinAshi/HeikinAshiPoint.ts:70]: Is this comment finished? Or is it indicating the below `forEach` loop?

File: ts/Series/HeikinAshi/HeikinAshiPoint.ts
Comment (raf18seb) [inline ts/Series/HeikinAshi/HeikinAshiPoint.ts:76]: Does it have to be `as any`? Can't we add a TS definition for `point.translated`?

File: ts/Series/HeikinAshi/HeikinAshiPoint.ts
Comment (raf18seb) [inline ts/Series/HeikinAshi/HeikinAshiPoint.ts:104]: Just a suggestion, this could be shorter
```js
public modifyFirstPointValue(): void {
    this.open = (this.open + this.high + this.low + this.close) / 4,
    this.close = (this.open + this.close) / 2;
}
```

File: ts/Series/HeikinAshi/HeikinAshiPoint.ts
Comment (raf18seb) [inline ts/Series/HeikinAshi/HeikinAshiPoint.ts:107]: ... point's values.

File: ts/Series/HeikinAshi/HeikinAshiPoint.ts
Comment (raf18seb) [inline ts/Series/HeikinAshi/HeikinAshiPoint.ts:130]: Same as above - could be shorter

**Review by raf18seb — CHANGES_REQUESTED** [review]

File: ts/Series/HeikinAshi/HeikinAshiPoint.ts
Comment (karolkolodziej) [reply ts/Series/HeikinAshi/HeikinAshiPoint.ts:130]: I'm using `newOpen` and `newClose` later.
I am afraid I won't be able to make it shorter.

File: ts/Series/HeikinAshi/HeikinAshiPoint.ts
Comment (karolkolodziej) [reply ts/Series/HeikinAshi/HeikinAshiPoint.ts:76]: It's not that easy to get rid of that `any`. 
Additional after some research, I see that we use that approach in other similar cases like this one so I guess for now it has to stay like this.

**Review by pawellysy — APPROVED** [review]
nothing to add! 👍

**Review by raf18seb — APPROVED** [review]
GJ, thanks!

**Review by pawelfus — CHANGES_REQUESTED** [review]
Looks great! Just minor polishing.

TODO: Decide if should be part of the Stock module, or a separate one

**Comment by karolkolodziej** [comment]
> TODO: Decide if should be part of the Stock module, or a separate one

I guess this is up to @TorsteinHonsi 
Could you please help us decide?

**Review by pawelfus — CHANGES_REQUESTED** [review]
Could you test this?
Try to use `series.modifyValue(point)`. I know you don't have access to the previous point with the current implementation, but adding an extra argument (`series.modifyValue(point, index)`) would let you access the previous point. Then you could remove the extra loop from translate()

**Comment by pawelfus** [comment]
> > TODO: Decide if should be part of the Stock module, or a separate one
> 
> I guess this is up to @TorsteinHonsi
> Could you please help us decide?

The longer I think about this, the more I'm leaning towards a separate module. Heikin Ashi and Hollow Candlestick series are just the two first series out of a few new ones I want to be implemented (Kagi, Renko, and Point&Figure). If every series will add 1kB+ to the package (I expect even more from the next three types), then it's the right moment to start releasing the Stock series as modules. I'm not sure how does this align with tree-shaking, though.

**Comment by TorsteinHonsi** [comment]
> it's the right moment to start releasing the Stock series as modules. I'm not sure how does this align with tree-shaking, though.

Agreed. To my knowledge, tree shaking shouldn't be affected by how the master files are set up.

> The only problem I see (with modularized series) is dataGrouping, which doesn't have a mechanism to easily extend it.

FYI @bre1470 . This should be taken into account when refactoring/migrating the data grouping.

**Review by TorsteinHonsi — APPROVED** [review]
Looks good to me, pending final approval by @pawelfus

File: ts/Extensions/DataGrouping.ts
Comment (pawelfus) [inline ts/Extensions/DataGrouping.ts:927]: Please, add a comment to move this to `HeikinAshiSeries.ts` when dataGrouping will support any form of extending with new types

**Review by pawelfus — CHANGES_REQUESTED** [review]

File: ts/Series/HeikinAshi/HeikinAshiSeries.ts
Comment (pawelfus) [inline ts/Series/HeikinAshi/HeikinAshiSeries.ts:157]: Why casting to `any`? The same below:
```
                const dataPoint: any = processedYData[i],
                    previousDataPoint: any = heikiashiData[i - 1];
```

File: ts/Series/HeikinAshi/HeikinAshiSeries.ts
Comment (pawelfus) [inline ts/Series/HeikinAshi/HeikinAshiSeries.ts:235]: This check seems to be unnecessary (`this.heikinashiData` is set on `init`)

**Review by pawelfus — CHANGES_REQUESTED** [review]
Looks great, just some minor polishing, thanks!

File: ts/Series/HeikinAshi/HeikinAshiSeries.ts
Comment (karolkolodziej) [reply ts/Series/HeikinAshi/HeikinAshiSeries.ts:157]: The `firstPoint` is an element from the `processedYData` so it might have a type like: `number|null|undefined`.
However, after removing the `undefined` from all definitions everything works as expected. 
(I guess this type might be omitted but I didn't want to make a mess in that PR.)

So my only concern was the `number|null` types. 
In calculation, adding `null` to something doesn't change the result, so I cast it to any to avoid checks or conversion from `null` to `0`.
Is there some easy hack to avoid that casting?

Comment (karolkolodziej) [reply ts/Series/HeikinAshi/HeikinAshiSeries.ts:157]: Comment about that is added below because previously I cast while invoking the method.

`this.modifyDataPoint(dataPoint, previousDataPoint);`

Comment (pawelfus) [reply ts/Series/HeikinAshi/HeikinAshiSeries.ts:157]: Okay, thanks for the explanation!

**Review by pawelfus — APPROVED** [review]
Great work! Thank you :)

**Review by pawellysy — APPROVED** [review]

**Review by raf18seb — APPROVED** [review]
GJ, thanks!

**Pull request merged**
